src/ngraph_assign_clusters.cc

```diff
@@ -83,23 +91,203 @@ namespace {
+      std::string src_cluster_dst_pred =
+          cluster_map[src_cluster_dst]->predicate_string;
+      if (pred_check != src_cluster_dst_pred) {
+        found_same_out_preds = false;
+        break;
+      }
+    }
+    // Cannot contract this edge
+    if (!found_same_out_preds) {
+      is_deadness_ok = false;
+      return Status::OK();
+    }
+  }
+
+  // Case src X, dst Y, X==Y
+  // Case src True, dst Y(Y can be True), all other output edges from src
+  // cluster have pred Y
+  // Ok to contract
+  is_deadness_ok = true;
+  return Status::OK();
 }
 
+// Some sanity checks for Node's cluster assignment wrt Deadness
+Status CheckNodeClusterAssignmentWRTDeadness(
+    Node* node, std::map<Node*, string>& nodes_predicate_map,
+    string& cluster_pred_string) {
+  if (nodes_predicate_map.find(node) == nodes_predicate_map.end()) {
+    return errors::Internal("Node ", node->name(), " [", node->type_string(),
+                            "]", " not found in predicate map");
+  }
+  std::string node_pred_string = nodes_predicate_map[node];
+
+  if (DeadnessAnalysis::IsControlFlowPredString(node_pred_string)) {
+    return errors::Internal(
+        "Node ", node->name(), " [", node->type_string(), "]",
+        " should not be clustered as it is a control flow op");
+  }
+
+  if (!DeadnessAnalysis::IsTruePredString(node_pred_string) &&
+      node_pred_string != cluster_pred_string) {
+    return errors::Internal(
+        "Node ", node->name(), " [", node->type_string(), "]", " Predicate : ",
+        node_pred_string,
+        "should not be clustered in cluster with pred_String ",
+        cluster_pred_string);
+  }
+  return Status::OK();
+}
+#endif
+
+// Merges src and dst clusters of the edge
+void MergeClusters(Edge* edge,
+                   std::map<Node*, std::shared_ptr<Cluster>>& cluster_map) {
+  Node* src = edge->src();
+  Node* dst = edge->dst();
+  int src_index = cluster_map[src]->index;
+  int dst_index = cluster_map[dst]->index;
+
+  // Merge dst cluster into src cluster
+  NGRAPH_VLOG(5) << "Contracting: " << src->name() << "[" << src->type_string()
+                 << " , " << edge->src_output() << "]@" << src_index << " -> "
+                 << dst->name() << "[" << dst->type_string() << " , "
+                 << edge->dst_input() << "]@" << dst_index;
+
+#if !defined(NGRAPH_TF_DISABLE_DEADNESS_CHECK)
+  string src_predicate = cluster_map[src]->predicate_string;
+  string dst_predicate = cluster_map[dst]->predicate_string;
+  NGRAPH_VLOG(5) << "Src pred: " << src_predicate
+                 << " ,Dst pred: " << dst_predicate;
+
+  std::string cluster_pred = GetMergedClusterPred(src_predicate, dst_predicate);
+
+  cluster_map[src]->predicate_string = cluster_pred;
+  // Update outgoing edges of the merged cluster
+  cluster_map[src]->outgoing_edges.insert(
+      cluster_map[dst]->outgoing_edges.begin(),
+      cluster_map[dst]->outgoing_edges.end());
+  cluster_map[src]->outgoing_edges.erase(edge);
+#endif
```

Comment:
maybe you have already done this: is it possible to run a small test/network to make sure that if NGRAPH_TF_DISABLE_DEADNESS_CHECK is set, then all the macros work as they should and we do not get compile errors or runtime errors.

Reply:
Yes I have done this test, before the PR